### DISCUSSION TITLE
spinner with invoke.run pty=True

### DISCUSSION THREAD

**Discussion by tompropst** [Textualize/rich#3177] [Q&A]
With the following code, you'll see that only every other `echo` produces a value on the screen. Am I missing something obvious or is there a fundamental conflict between the spinner and the `invoke` `pty` option?

```python
import time

from invoke import task
from rich.console import Console


@task()
def main(ctx):                                   
    console = Console()
    console.print("Starting")                    
    with console.status("Running"):              
        for i in range(0, 10):                   
            ctx.run(f"echo {i}", pty=bool(i % 2))
            time.sleep(1)
```

Here's a sample run:

![image](https://github.com/Textualize/rich/assets/4935266/0c6eebe8-8f26-4e55-beb9-d4b0e00136e2)

**Comment by willmcgugan** [comment]
I'm assuming that `invoke` runs the `echo` process. As its writing directly to stdout, Rich isn't able to update the spinner. You will need to capture the output from the process and print it with the console for this to work.

**Comment by tompropst** [comment]
Thanks for the quick response @willmcgugan! I should be able to process the command output manually and pass to the console.

There is an option in invoke to send the output to a "file like" object instead. Is there by chance any mechanism to treat a console like a file handle?

From the [invoke docs](https://docs.pyinvoke.org/en/stable/api/runners.html):

> out_stream – A file-like stream object to which the subprocess’ standard output should be written. If None (the default), sys.stdout will be used.
>
> pty bool – By default, run connects directly to the invoked process and reads its stdout/stderr streams. Some programs will buffer (or even behave) differently in this situation compared to using an actual terminal or pseudoterminal (pty). To use a pty instead of the default behavior, specify pty=True.

I should have mentioned that I'm only trying to use the `pty` option so that I can get the color output of the process run by invoke. I may be able to get it to look as good or better by reprinting with a rich console.
